File: Engine/Core/command_builder.py
from __future__ import annotations
import platform, sys
from pathlib import Path
from typing import Dict, Any, List
from Engine.Utils.printer import print, Colours
from Engine.Core.placeholders import resolve_placeholders


class CommandBuilder:
	def __init__(self, root_path: Path):
		self.root_path = root_path


	def _get_executable_for_operation(self, op: Dict[str, Any]) -> str:
		script_type = op.get("script_type", "python").lower()
		if script_type == "python":
			is_win64 = sys.platform == 'win32' and platform.machine().endswith('64')
			if is_win64:
				local = self.root_path / "runtime" / "python3" / "python.exe"
				if local.is_file():
					print(colour=Colours.CYAN, message=f"Using local Python runtime: {local}")
					return str(local)
				print(colour=Colours.YELLOW, message="Local Python runtime not found, checking system PATH.")
			print(colour=Colours.CYAN, message="Using 'python' from system PATH.")
			return "python"
		print(colour=Colours.CYAN, message=f"Using '{script_type}' from system PATH.")
		return script_type


	def build(self, current_game: str, games: Dict[str, Any], engine_config: Dict[str, Any], op: Dict[str, Any], prompt_answers: Dict[str, Any]) -> List[str]:
		if not current_game:
			raise ValueError("No game has been loaded.")
		ctx = dict(engine_config)
		ctx["Game"] = {"RootPath": str(games[current_game]["game_root"]), "Name": current_game}
		executable = self._get_executable_for_operation(op)
		script_path = op.get("script")
		if not script_path:
			return []
		script_path = resolve_placeholders(script_path, ctx)
		parts = [executable, script_path]
		static_args = op.get("args", [])
		resolved_args = resolve_placeholders(static_args, ctx)
		parts.extend([str(a) for a in resolved_args if a is not None])
		for prompt in op.get("prompts", []):
			name = prompt["Name"]
			ans = prompt_answers.get(name)
			if "condition" in prompt and not prompt_answers.get(prompt["condition"]):
				continue
			if prompt["type"] == "confirm" and ans and "cli_arg" in prompt:
				parts.append(prompt["cli_arg"])
			elif prompt["type"] == "checkbox" and ans and "cli_prefix" in prompt:
				parts.append(prompt["cli_prefix"])
				parts.extend([str(a) for a in ans])
			elif prompt["type"] == "text" and ans and str(ans).strip():
				v = str(ans).strip()
				if "cli_arg_prefix" in prompt:
					parts.append(prompt["cli_arg_prefix"]); parts.append(v)
				elif "cli_arg" in prompt:
					parts.append(prompt["cli_arg"]); parts.append(v)
		return parts

File: EngineNet/Core/Sys/CommandBuilder.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Diagnostics;

namespace EngineNet.Core.Sys;

/// <summary>
/// Constructs command-line invocations from an operation definition and context.
/// Responsible for resolving placeholders and mapping prompts/answers to CLI args.
/// </summary>
public sealed class CommandBuilder {
    private readonly String _rootPath;
    /// <summary>
    /// Initializes a new instance of <see cref="CommandBuilder"/>.
    /// </summary>
    /// <param name="rootPath">Working root of the engine; may be used for relative resolution.</param>
    public CommandBuilder(String rootPath) {
        _rootPath = rootPath;
    }

    private String GetExecutableForOperation(IDictionary<String, Object?> op) {
        String scriptType = (op.TryGetValue("script_type", out Object? st) ? st?.ToString() : null)?.ToLowerInvariant() ?? "python";
        if (scriptType == "python") {
            Boolean isWin = OperatingSystem.IsWindows();
            return isWin ? "python" : "python3";
        }
        return scriptType; // just returns script type as executable name
    }

    /// <summary>
    /// Build a process invocation for the given operation.
    /// </summary>
    /// <param name="currentGame">Canonical game/module id currently selected.</param>
    /// <param name="games">Map of games with metadata; must contain <paramref name="currentGame"/>.</param>
    /// <param name="engineConfig">Engine configuration dictionary exposed to placeholder resolution.</param>
    /// <param name="op">The operation object (must include at least <c>script</c>, optional <c>script_type</c>, <c>args</c>, and <c>prompts</c>).</param>
    /// <param name="promptAnswers">Mutable dictionary of prompt answers used to construct arguments; defaults are filled if missing.</param>
    /// <returns>A list of strings suitable for <see cref="ProcessStartInfo"/>: [exe, arg1, ...]. Empty if no script.</returns>
    /// <exception cref="ArgumentException">Thrown when <paramref name="currentGame"/> is empty.</exception>
    /// <exception cref="KeyNotFoundException">Thrown when the game is unknown.</exception>
    public List<String> Build(
        String currentGame,
        IDictionary<String, Object?> games,
        IDictionary<String, Object?> engineConfig,
        IDictionary<String, Object?> op,
        IDictionary<String, Object?> promptAnswers) {
        if (String.IsNullOrWhiteSpace(currentGame)) {
            throw new ArgumentException("No game has been loaded.", nameof(currentGame));
        }

        // Build context = engineConfig + built-ins + Game{ RootPath, Name }
        Dictionary<String, Object?> ctx = new(engineConfig, StringComparer.OrdinalIgnoreCase);

        if (!games.TryGetValue(currentGame, out Object? g) || g is not IDictionary<String, Object?> gdict) {
            throw new KeyNotFoundException($"Unknown game '{currentGame}'.");
        }

        String gameRoot = gdict.TryGetValue("game_root", out Object? gr) ? gr?.ToString() ?? "" : "";
        // Inject built-in, non-dotted placeholders for convenience
        // Examples:
        //  - {{Game_Root}} -> path to the active module/game
        //  - {{Project_Root}} -> engine project root folder
        //  - {{Registry_Root}} -> RemakeRegistry folder under project root
        ctx["Game_Root"] = gameRoot;
        ctx["Project_Root"] = _rootPath;
        ctx["Registry_Root"] = Path.Combine(_rootPath, "RemakeRegistry");

        ctx["Game"] = new Dictionary<String, Object?> {
            ["RootPath"] = gameRoot,
            ["Name"] = currentGame,
        };

        // Back-compat: ensure dynamic RemakeEngine.Config.module_path and project_path exist
        if (!ctx.TryGetValue("RemakeEngine", out Object? re) || re is not IDictionary<String, Object?> reDict) {
            ctx["RemakeEngine"] = reDict = new Dictionary<String, Object?>(StringComparer.OrdinalIgnoreCase);
        }

        if (!reDict.TryGetValue("Config", out Object? cfg) || cfg is not IDictionary<String, Object?> cfgDict) {
            reDict["Config"] = cfgDict = new Dictionary<String, Object?>(StringComparer.OrdinalIgnoreCase);
        }

        cfgDict["module_path"] = gameRoot;
        cfgDict["project_path"] = _rootPath;

        // Merge module-specific placeholders from config.toml (if present)
        try {
            String cfgPath = Path.Combine(gameRoot, "config.toml");
            if (!String.IsNullOrWhiteSpace(gameRoot) && File.Exists(cfgPath)) {
                Dictionary<String, Object?> fromToml = Tools.SimpleToml.ReadPlaceholdersFile(cfgPath);
                foreach (KeyValuePair<String, Object?> kv in fromToml) {
                    if (!ctx.ContainsKey(kv.Key)) {
                        ctx[kv.Key] = kv.Value;
                    }
                }
            }
        } catch {
            // non-fatal: ignore bad/missing config
        }

        String executable = GetExecutableForOperation(op);
        if (!op.TryGetValue("script", out Object? scriptObj)) {
            return [];
        }

        String scriptPath = Placeholders.Resolve(scriptObj, ctx)?.ToString() ?? String.Empty;
        List<String> parts = [executable, scriptPath];

        if (op.TryGetValue("args", out Object? argsObj) && argsObj is IList<Object?> aList) {
            IList<Object?> resolved = (IList<Object?>)(Placeholders.Resolve(aList, ctx) ?? new List<Object?>());
            foreach (Object? a in resolved) {
                if (a is not null) {
                    parts.Add(a.ToString()!);
                }
            }
        }

        if (op.TryGetValue("prompts", out Object? promptsObj) && promptsObj is IList<Object?> prompts) {
            // First pass: seed promptAnswers with defaults so conditions can evaluate
            foreach (Object? p in prompts) {
                if (p is not IDictionary<String, Object?> prompt) {
                    continue;
                }

                String name = prompt.TryGetValue("Name", out Object? n) ? n?.ToString() ?? "" : "";
                String type = prompt.TryGetValue("type", out Object? t) ? t?.ToString() ?? "" : "";
                if (String.IsNullOrEmpty(name) || String.IsNullOrEmpty(type)) {
                    continue;
                }

                if (!promptAnswers.ContainsKey(name) && prompt.TryGetValue("default", out Object? defVal)) {
                    promptAnswers[name] = defVal;
                }
            }

            // Second pass: map answers to CLI args respecting conditions
            foreach (Object? p in prompts) {
                if (p is not IDictionary<String, Object?> prompt) {
                    continue;
                }

                String name = prompt.TryGetValue("Name", out Object? n) ? n?.ToString() ?? "" : "";
                String type = prompt.TryGetValue("type", out Object? t) ? t?.ToString() ?? "" : "";

                if (String.IsNullOrEmpty(name) || String.IsNullOrEmpty(type)) {
                    continue;
                }

                if (prompt.TryGetValue("condition", out Object? cond) && cond is String condName) {
                    if (!promptAnswers.TryGetValue(condName, out Object? condVal) || condVal is not Boolean b || !b) {
                        continue;
                    }
                }

                _ = promptAnswers.TryGetValue(name, out Object? ans);

                switch (type) {
                    case "confirm":
                        if (ans is Boolean cb && cb && prompt.TryGetValue("cli_arg", out Object? cli) && cli is String s1) {
                            parts.Add(s1);
                        }

                        break;
                    case "checkbox":
                        if (ans is IList<Object?> items && prompt.TryGetValue("cli_prefix", out Object? pref) && pref is String sp) {
                            parts.Add(sp);
                            foreach (Object? it in items) {
                                if (it is not null) {
                                    parts.Add(it.ToString()!);
                                }
                            }
                        }
                        break;
                    case "text":
                        String? v = ans?.ToString();
                        if (!String.IsNullOrWhiteSpace(v)) {
                            if (prompt.TryGetValue("cli_arg_prefix", out Object? ap) && ap is String apx) {
                                parts.Add(apx);
                                parts.Add(v);
                            } else if (prompt.TryGetValue("cli_arg", out Object? ca) && ca is String cas) {
                                parts.Add(cas);
                                parts.Add(v);
                            }
                        }
                        break;
                }
            }
        }

        return parts;
    }
}
